File: app/src/main/java/de/passwordvault/model/storage/csv/CsvBuilder.java
package de.passwordvault.model.storage.csv;

import androidx.annotation.NonNull;


/**
 * Class models a builder which can build CSV.
 *
 * @author  Christian-2003
 * @version 3.3.0
 */
public class CsvBuilder extends CsvConfiguration {

    /**
     * Attribute stores the StringBuilder which is used internally to build the CSV.
     */
    private final StringBuilder csv;

    /**
     * Attribute stores whether {@link #csv} already contains some content.
     */
    private boolean containsContent;


    /**
     * Constructor instantiates a new {@link CsvBuilder}-instance.
     */
    public CsvBuilder() {
        csv = new StringBuilder();
        containsContent = false;
    }


    /**
     * Method appends the passed argument to the generated CSV.
     *
     * @param arg   Value to be appended to the CSV.
     */
    public void append(int arg) {
        if (containsContent) {
            csv.append(COLUMN_DIVIDER);
        }
        csv.append(arg);
        containsContent = true;
    }

    /**
     * Method appends the passed argument to the generated CSV.
     *
     * @param arg   Value to be appended to the CSV.
     */
    public void append(byte arg) {
        if (containsContent) {
            csv.append(COLUMN_DIVIDER);
        }
        csv.append(arg);
        containsContent = true;
    }

    /**
     * Method appends the passed argument to the generated CSV.
     *
     * @param arg   Value to be appended to the CSV.
     */
    public void append(short arg) {
        if (containsContent) {
            csv.append(COLUMN_DIVIDER);
        }
        csv.append(arg);
        containsContent = true;
    }

    /**
     * Method appends the passed argument to the generated CSV.
     *
     * @param arg   Value to be appended to the CSV.
     */
    public void append(long arg) {
        if (containsContent) {
            csv.append(COLUMN_DIVIDER);
        }
        csv.append(arg);
        containsContent = true;
    }

    /**
     * Method appends the passed argument to the generated CSV.
     *
     * @param arg   Value to be appended to the CSV.
     */
    public void append(float arg) {
        if (containsContent) {
            csv.append(COLUMN_DIVIDER);
        }
        csv.append(arg);
        containsContent = true;
    }

    /**
     * Method appends the passed argument to the generated CSV.
     *
     * @param arg   Value to be appended to the CSV.
     */
    public void append(double arg) {
        if (containsContent) {
            csv.append(COLUMN_DIVIDER);
        }
        csv.append(arg);
        containsContent = true;
    }

    /**
     * Method appends the passed argument to the generated CSV.
     *
     * @param arg   Value to be appended to the CSV.
     */
    public void append(char arg) {
        if (containsContent) {
            csv.append(COLUMN_DIVIDER);
        }
        csv.append(arg);
        containsContent = true;
    }

    /**
     * Method appends the passed argument to the generated CSV.
     *
     * @param arg   Value to be appended to the CSV.
     */
    public void append(boolean arg) {
        if (containsContent) {
            csv.append(COLUMN_DIVIDER);
        }
        csv.append(arg);
        containsContent = true;
    }

    /**
     * Method appends the passed argument to the generated CSV. Passing {@code null} is allowed,
     * but not recommended.
     *
     * @param arg   Value to be appended to the CSV.
     */
    public void append(String arg) {
        if (containsContent) {
            csv.append(COLUMN_DIVIDER);
        }
        if (arg == null) {
            csv.append('\0');
            containsContent = true;
            return;
        }
        boolean containsSpecialCharacters = false;
        StringBuilder string = new StringBuilder(); //I'm funny I know...
        for (int i = 0; i < arg.length(); i++) {
            char currentChar = arg.charAt(i);
            switch (currentChar) {
                case STRING_SEPARATOR:
                    string.append("\\"); //Escape string separator for CSV structure.
                    string.append(currentChar);
                    containsSpecialCharacters = true;
                    break;
                case COLUMN_DIVIDER:
                    string.append(currentChar);
                    containsSpecialCharacters = true;
                    break;
                case ROW_DIVIDER:
                    string.append("\\n"); //Escape row divider.
                    containsSpecialCharacters = true;
                    break;
                default:
                    string.append(currentChar);
            }
        }
        if (containsSpecialCharacters) {
            csv.append(STRING_SEPARATOR);
        }
        csv.append(string);
        if (containsSpecialCharacters) {
            csv.append(STRING_SEPARATOR);
        }
        containsContent = true;
    }

    /**
     * Method appends the passed argument to the generated CSV. Passing {@code null} is allowed,
     * but not recommended.
     *
     * @param arg   Value to be appended to the CSV.
     */
    public void append(Object arg) {
        if (arg == null) {
            if (containsContent) {
                append(COLUMN_DIVIDER);
            }
            append('\0');
            containsContent = true;
        }
        else {
            append(arg.toString());
        }
    }


    /**
     * Method appends a line feed to the generated CSV.
     */
    public void newLine() {
        csv.append(ROW_DIVIDER);
        containsContent = false;
    }


    /**
     * Method converts the generated CSV into a string.
     *
     * @return  Generated CSV.
     */
    @NonNull
    @Override
    public String toString() {
        return csv.toString();
    }

}

File: core/src/main/java/de/christian2003/core/conversion/csv/CsvBuilder.kt
package de.christian2003.core.conversion.csv


/**
 * Class implements a CSV builder that can generate CSV content.
 *
 * @param config    Config determining the syntax of the CSV generated.
 * @author          Christian-2003
 * @since           3.8.0
 */
class CsvBuilder(

    /**
     * Config to use when creating CSV files.
     */
    private val config: CsvConfig = CsvConfig()

) {

    /**
     * Stores the CSV that has been generated so far.
     */
    private val csv: StringBuilder = StringBuilder()

    /**
     * Stores whether the current line in CSV contains any content.
     */
    private var isCurrentLineEmpty: Boolean = true


    /**
     * Appends the argument specified to the current CSV row.
     *
     * @param arg   Value to append to the current CSV row.
     * @return      This CSV builder.
     */
    fun append(arg: Int): CsvBuilder {
        if (!isCurrentLineEmpty) {
            csv.append(config.columnDivider)
        }
        csv.append(arg)
        isCurrentLineEmpty = false
        return this
    }


    /**
     * Appends the argument specified to the current CSV row.
     *
     * @param arg   Value to append to the current CSV row.
     * @return      This CSV builder.
     */
    fun append(arg: Byte): CsvBuilder {
        if (!isCurrentLineEmpty) {
            csv.append(config.columnDivider)
        }
        csv.append(arg)
        isCurrentLineEmpty = false
        return this
    }



    /**
     * Appends the argument specified to the current CSV row.
     *
     * @param arg   Value to append to the current CSV row.
     * @return      This CSV builder.
     */
    fun append(arg: Long): CsvBuilder {
        if (!isCurrentLineEmpty) {
            csv.append(config.columnDivider)
        }
        csv.append(arg)
        isCurrentLineEmpty = false
        return this
    }


    /**
     * Appends the argument specified to the current CSV row.
     *
     * @param arg   Value to append to the current CSV row.
     * @return      This CSV builder.
     */
    fun append(arg: Float): CsvBuilder {
        if (!isCurrentLineEmpty) {
            csv.append(config.columnDivider)
        }
        csv.append(arg)
        isCurrentLineEmpty = false
        return this
    }


    /**
     * Appends the argument specified to the current CSV row.
     *
     * @param arg   Value to append to the current CSV row.
     * @return      This CSV builder.
     */
    fun append(arg: Double): CsvBuilder {
        if (!isCurrentLineEmpty) {
            csv.append(config.columnDivider)
        }
        csv.append(arg)
        isCurrentLineEmpty = false
        return this
    }


    /**
     * Appends the argument specified to the current CSV row.
     *
     * @param arg   Value to append to the current CSV row.
     * @return      This CSV builder.
     */
    fun append(arg: Char): CsvBuilder {
        if (!isCurrentLineEmpty) {
            csv.append(config.columnDivider)
        }
        csv.append(arg)
        isCurrentLineEmpty = false
        return this
    }


    /**
     * Appends the argument specified to the current CSV row.
     *
     * @param arg   Value to append to the current CSV row.
     * @return      This CSV builder.
     */
    fun append(arg: Boolean): CsvBuilder {
        if (!isCurrentLineEmpty) {
            csv.append(config.columnDivider)
        }
        csv.append(arg)
        isCurrentLineEmpty = false
        return this
    }


    /**
     * Appends the string specified to the current CSV row. If the string contains any special
     * characters that are used with the CSV syntax, it is escaped using a string separator.
     *
     * @param arg   String to append to the current CSV row.
     * @return      This CSV builder.
     */
    fun append(arg: String): CsvBuilder {
        if (!isCurrentLineEmpty) {
            csv.append(config.columnDivider)
        }

        var containsSpecialCharacters = false
        val builder = StringBuilder()

        arg.forEach { char ->
            when(char) {
                config.columnDivider -> {
                    builder.append(char)
                    containsSpecialCharacters = true
                }
                config.rowDivider -> {
                    builder.append("\\n")
                    containsSpecialCharacters = true
                }
                config.stringSeparator -> {
                    builder.append("\\")
                    builder.append(char)
                    containsSpecialCharacters = true
                }
                else -> {
                    builder.append(char)
                }
            }
        }

        if (containsSpecialCharacters) {
            csv.append(config.stringSeparator)
        }
        csv.append(builder)
        if (containsSpecialCharacters) {
            csv.append(config.stringSeparator)
        }

        isCurrentLineEmpty = false
        return this
    }


    /**
     * Adds a new line to the CSV.
     *
     * @return  This CSV builder.
     */
    fun newLine(): CsvBuilder {
        csv.append(config.rowDivider)
        isCurrentLineEmpty = true
        return this
    }


    /**
     * Converts the CSV built so far into a string.
     *
     * @return  CSV as string.
     */
    override fun toString(): String {
        return csv.toString()
    }

}
